src/gui/instrument/EnvelopeGraph.cpp

```diff
@@ -54,102 +56,208 @@ void EnvelopeGraph::modelChanged()
+	const float holdPercentage = hold / params->getHoldModel().maxValue();
+
 	const float decay = params->getDecayModel().value();
+	const float decayPercentage = decay / params->getDecayModel().maxValue();
+
 	const float sustain = params->getSustainModel().value();
+
 	const float release = params->getReleaseModel().value();
+	const float releasePercentage = release / params->getReleaseModel().maxValue();
+
+	// The margin to the left and right so that we do not clip too much of the lines and markers
+	const float margin = 2.0;
+	const float availableWidth = width() - margin * 2;
 
-	const float gray_amount = 1.0f - fabsf(amount);
+	// Now determine the maximum width for one segment according to the scaling setting.
+	// The different scalings use different means to compute the maximum available width per segment.
+	const auto computeMaximumSegmentWidthAbsolute = [&]() -> float
+	{
+		return availableWidth / 5.;
+	};
 
-	p.setPen(QPen(QColor(static_cast<int>(96 * gray_amount),
-				static_cast<int>(255 - 159 * gray_amount),
-				static_cast<int>(128 - 32 * gray_amount)),
-									2));
+	const auto computeMaximumSegmentWidthRelative = [&]() -> float
+	{
+		const float sumOfSegments = predelayPercentage + attackPercentage + holdPercentage + decayPercentage + releasePercentage;
 
-	const QColor end_points_color(0x99, 0xAF, 0xFF);
-	const QColor end_points_bg_color(0, 0, 2);
+		if (sumOfSegments != 0.)
+		{
+			return availableWidth / sumOfSegments;
+		}
+		else
+		{
+			return computeMaximumSegmentWidthAbsolute();
+		}
+	};
 
-	const int y_base = m_envGraph.height() - 3;
-	const int avail_height = m_envGraph.height() - 6;
-	
-	int x1 = static_cast<int>(predelay * TIME_UNIT_WIDTH);
-	int x2 = x1 + static_cast<int>(attack * TIME_UNIT_WIDTH);
-	int x3 = x2 + static_cast<int>(hold * TIME_UNIT_WIDTH);
-	int x4 = x3 + static_cast<int>((decay * (1 - sustain)) * TIME_UNIT_WIDTH);
-	int x5 = x4 + static_cast<int>(release * TIME_UNIT_WIDTH);
+	const auto computeMaximumSegmentWidthDynamic = [&]() -> float
+	{
+		const float sumOfSegments = predelayPercentage + attackPercentage + holdPercentage + decayPercentage + releasePercentage;
 
-	if (x5 > 174)
+		float preliminarySegmentWidth = 80. / 182. * availableWidth;
+		
+		const float neededWidth = sumOfSegments * preliminarySegmentWidth;
+		
+		if (neededWidth > availableWidth && sumOfSegments != 0.)
+		{
+			return computeMaximumSegmentWidthRelative();
+		}
+		else
+		{
+			return preliminarySegmentWidth;
+		}
+	};
+
+	// This is the maximum width that each of the five segments (DAHDR) can occupy.
+	float maximumSegmentWidth;
+
+	switch (m_scaling)
 	{
-		x1 = (x1 * 174) / x5;
-		x2 = (x2 * 174) / x5;
-		x3 = (x3 * 174) / x5;
-		x4 = (x4 * 174) / x5;
-		x5 = (x5 * 174) / x5;
+		case ScalingMode::Absolute:
+			maximumSegmentWidth = computeMaximumSegmentWidthAbsolute();
+			break;
+		case ScalingMode::Relative:
+		{
+			maximumSegmentWidth = computeMaximumSegmentWidthRelative();
+			break;
+		}
```

Comment:
Since we're not making any temporary variables we don't need to scope block here and below. `Absolute` above is good.

Reply:
Adjusted with commit af4672b2094.